File: CS/DiagramTwoDataSources/EdgesDataTableDataProvider.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Linq;
using System.Web;

namespace DiagramTwoDataSources
{
    public static class EdgesDataTableDataProvider
    {
        private static DataTable Edges
        {
            get
            {
                var dataTable = HttpContext.Current.Session["DiagramDataTableEdges"] as DataTable;
                if (dataTable == null)
                {
                    dataTable = new DataTable();

                    dataTable.Columns.Add("ID", typeof(int));
                    dataTable.Columns.Add("Text", typeof(string));
                    dataTable.Columns.Add("FromID", typeof(int));
                    dataTable.Columns.Add("ToID", typeof(int));

                    dataTable.PrimaryKey = new DataColumn[] { dataTable.Columns["ID"] };

                    dataTable.Rows.Add(1, null, 1, 2);
                    dataTable.Rows.Add(2, null, 2, 3);
                    dataTable.Rows.Add(3, "No", 3, 5);
                    dataTable.Rows.Add(4, null, 5, 2);
                    dataTable.Rows.Add(5, null, 8, 4);
                    dataTable.Rows.Add(6, string.Empty, 4, 6);
                    dataTable.Rows.Add(9, string.Empty, 4, 7);
                    dataTable.Rows.Add(10, "Yes", 3, 8);
                    dataTable.Rows.Add(11, "Need developer assistance?", 8, 9);
                    dataTable.Rows.Add(12, null, 9, 4);

                    HttpContext.Current.Session["DiagramDataTableEdges"] = dataTable;
                }
                return dataTable;
            }
        }

        [DataObjectMethod(DataObjectMethodType.Select, true)]
        public static DataTable GetEdges()
        {
            return Edges;
        }

        [DataObjectMethod(DataObjectMethodType.Insert, true)]
        public static int InsertEdge(string text, int fromId, int toId)
        {
            int newId = (int)Edges.Compute("max([ID])", string.Empty) + 1;
            Edges.Rows.Add(newId, text, fromId, toId);
            return newId;
        }

        [DataObjectMethod(DataObjectMethodType.Update, true)]
        public static void UpdateEdge(int id, string text, int fromId, int toId)
        {
            var edgeToUpdate = Edges.Rows.Find(id);
            edgeToUpdate["Text"] = text;
            edgeToUpdate["FromID"] = fromId;
            edgeToUpdate["ToID"] = toId;
        }

        [DataObjectMethod(DataObjectMethodType.Delete, true)]
        public static void DeleteEdge(int id)
        {
            DataRow dr = Edges.Rows.Find(id);
            Edges.Rows.Remove(dr);
        }
    }
}

File: VB/DiagramTwoDataSources/EdgesDataTableDataProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Linq
Imports System.Web

Namespace DiagramTwoDataSources
	Public Module EdgesDataTableDataProvider
		Private ReadOnly Property Edges() As DataTable
			Get
				Dim dataTable = TryCast(HttpContext.Current.Session("DiagramDataTableEdges"), DataTable)
				If dataTable Is Nothing Then
					dataTable = New DataTable()

					dataTable.Columns.Add("ID", GetType(Integer))
					dataTable.Columns.Add("Text", GetType(String))
					dataTable.Columns.Add("FromID", GetType(Integer))
					dataTable.Columns.Add("ToID", GetType(Integer))

					dataTable.PrimaryKey = New DataColumn() { dataTable.Columns("ID") }

					dataTable.Rows.Add(1, Nothing, 1, 2)
					dataTable.Rows.Add(2, Nothing, 2, 3)
					dataTable.Rows.Add(3, "No", 3, 5)
					dataTable.Rows.Add(4, Nothing, 5, 2)
					dataTable.Rows.Add(5, Nothing, 8, 4)
					dataTable.Rows.Add(6, String.Empty, 4, 6)
					dataTable.Rows.Add(9, String.Empty, 4, 7)
					dataTable.Rows.Add(10, "Yes", 3, 8)
					dataTable.Rows.Add(11, "Need developer assistance?", 8, 9)
					dataTable.Rows.Add(12, Nothing, 9, 4)

					HttpContext.Current.Session("DiagramDataTableEdges") = dataTable
				End If
				Return dataTable
			End Get
		End Property

		<DataObjectMethod(DataObjectMethodType.Select, True)>
		Public Function GetEdges() As DataTable
			Return Edges
		End Function

		<DataObjectMethod(DataObjectMethodType.Insert, True)>
		Public Function InsertEdge(ByVal text As String, ByVal fromId As Integer, ByVal toId As Integer) As Integer
			Dim newId As Integer = DirectCast(Edges.Compute("max([ID])", String.Empty), Integer) + 1
			Edges.Rows.Add(newId, text, fromId, toId)
			Return newId
		End Function

		<DataObjectMethod(DataObjectMethodType.Update, True)>
		Public Sub UpdateEdge(ByVal id As Integer, ByVal text As String, ByVal fromId As Integer, ByVal toId As Integer)
			Dim edgeToUpdate = Edges.Rows.Find(id)
			edgeToUpdate("Text") = text
			edgeToUpdate("FromID") = fromId
			edgeToUpdate("ToID") = toId
		End Sub

		<DataObjectMethod(DataObjectMethodType.Delete, True)>
		Public Sub DeleteEdge(ByVal id As Integer)
			Dim dr As DataRow = Edges.Rows.Find(id)
			Edges.Rows.Remove(dr)
		End Sub
	End Module
End Namespace